File: SupplierPortal.Web/Modules/Market/Offer/OfferForm.cs
using Serenity.ComponentModel;
using SupplierPortal.Market.Enum;

namespace SupplierPortal.Market.Forms;

[FormScript("Market.Offer")]
[BasedOnRow(typeof(OfferRow), CheckNames = true)]
public class OfferForm
{
    [OfferSupplierEditor]
    public List<OfferSupplierRow> CompanyName { get; set; }
    [Visible(false)]
    public State State { get; set; }
    //[ReadOnly(true)]
    //public string Email { get; set; }
    [Width(100)]
    public DateTime OfferDate { get; set; }
    [Width(100)]
    public DateTime ValueDate { get; set; }

    [DisplayName("Detail"), OfferDetailEditor, IgnoreName]
    public List<OfferDetailRow> Detail { get; set; }
}

File: SupplierPortal.Web/Modules/ServerTypes/Market/OfferForm.ts
import { OfferDetailEditor } from "@/Market/OfferDetail/OfferDetailEditor";
import { OfferSupplierEditor } from "@/Market/OfferSupplier/OfferSupplierEditor";
import { EnumEditor, DateEditor, PrefixedContext, initFormType } from "@serenity-is/corelib";
import { State } from "./Enum.State";

export interface OfferForm {
    CompanyName: OfferSupplierEditor;
    State: EnumEditor;
    OfferDate: DateEditor;
    ValueDate: DateEditor;
    Detail: OfferDetailEditor;
}

export class OfferForm extends PrefixedContext {
    static readonly formKey = 'Market.Offer';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!OfferForm.init)  {
            OfferForm.init = true;

            var w0 = OfferSupplierEditor;
            var w1 = EnumEditor;
            var w2 = DateEditor;
            var w3 = OfferDetailEditor;

            initFormType(OfferForm, [
                'CompanyName', w0,
                'State', w1,
                'OfferDate', w2,
                'ValueDate', w2,
                'Detail', w3
            ]);
        }
    }
}

[State]; // referenced types